feat(metrics): create new metrics_counters aggregate table with updated granularity …

onewland opened the pull request:
…and use-case ID column

fpacifici commented on snuba/migrations/snuba_migrations/metrics/0025_metrics_counters_aggregate_v2.py:
```diff
@@ -0,0 +1,99 @@
+                columns=self.aggregated_cols,
+                engine=table_engines.AggregatingMergeTree(
+                    storage_set=StorageSetKey.METRICS,
+                    order_by="(org_id, project_id, metric_id, granularity, timestamp, tags.key, tags.value)",
```
We need to separate the ORDER BY key from the Primary key

Like this 
```
PRIMARY KEY (org_id, project_id, metric_id, granularity, timestamp)
ORDER BY (org_id, project_id, metric_id, granularity, timestamp, tags.key, tags.value)
```

This ensures we store value uniquely by ORDER BY and sorted that way but we include only `org_id, project_id, metric_id, granularity, timestamp` in the index, which reduces the index size.

Also, please add the use_case_id to the ORDER BY key as well (either before or after tags) but not to the PRIMARY KEY. The partition key should be enough to prevent Clickhosue from merging rows of different retentions but we never know.

fpacifici commented on snuba/migrations/snuba_migrations/metrics/0025_metrics_counters_aggregate_v2.py:
```diff
@@ -0,0 +1,99 @@
+    blocking = False
+    table_name = "metrics_counters_v2_local"
+    dist_table_name = "metrics_counters_v2_dist"
+    granularity = "1024"
```
Let's 2048 first ?

fpacifici commented on snuba/migrations/snuba_migrations/metrics/0025_metrics_counters_aggregate_v2.py:
```diff
@@ -0,0 +1,99 @@
+    def forwards_local(self) -> Sequence[operations.SqlOperation]:
+
+        return [
+            operations.CreateTable(
```
This is missing the TTL.

fpacifici reviewed (changes requested):
A couple of changes. Please see comments inline

onewland commented:
New SQL:
```
Local operations:
CREATE TABLE IF NOT EXISTS metrics_counters_v2_local (use_case_id LowCardinality(String), org_id UInt64, project_id UInt64, metric_id UInt64, granularity UInt32, tags Nested(key UInt64, value UInt64), timestamp DateTime, retention_days UInt16, value AggregateFunction(sum, Float64)) ENGINE AggregatingMergeTree()  PRIMARY KEY (org_id, project_id, metric_id, granularity, timestamp) ORDER BY (org_id, project_id, metric_id, granularity, timestamp, use_case_id, tags.key, tags.value) PARTITION BY (retention_days, toMonday(timestamp)) TTL timestamp + toIntervalDay(retention_days) SETTINGS index_granularity=2048;
ALTER TABLE metrics_counters_v2_local ADD COLUMN IF NOT EXISTS _tags_hash Array(UInt64) MATERIALIZED arrayMap((k, v) -> cityHash64(concat(toString(k), '=', toString(v))), tags.key, tags.value) AFTER tags.value;
ALTER TABLE metrics_counters_v2_local ADD INDEX IF NOT EXISTS bf_tags_hash _tags_hash TYPE bloom_filter() GRANULARITY 1;
ALTER TABLE metrics_counters_v2_local ADD INDEX IF NOT EXISTS bf_tags_key_hash tags.key TYPE bloom_filter() GRANULARITY 1;
```

onewland commented:
Addressed all comments AFAIK

fpacifici commented on snuba/migrations/snuba_migrations/metrics/0025_metrics_counters_aggregate_v2.py:
```diff
@@ -0,0 +1,100 @@
+                columns=self.aggregated_cols,
+                engine=table_engines.AggregatingMergeTree(
+                    storage_set=StorageSetKey.METRICS,
+                    order_by="(org_id, project_id, metric_id, granularity, timestamp, use_case_id, tags.key, tags.value)",
```
retention is still missing.
I thought `use_case_id` would be the first in the key. 
If we start using use_case_id we will always filter by that, so we should put it both in the order by and the primary key. 

The fields we have in the order by but not in the primary key are not part of the index, so we do not get any advantage there

onewland replied:
I'm fine with use_case_id being part of the `primary key` and `order by`, but was this comment above incorrect?

> Also, please add the use_case_id to the ORDER BY key as well (either before or after tags) but not to the PRIMARY KEY. The partition key should be enough to prevent Clickhosue from merging rows of different retentions but we never know.

And can you explain in more depth what `retention is still missing.` means? You want `retention_days` in the ORDER BY so that rows with different timestamps can't be merged (even though they should always be partitioned separately)?

fpacifici commented:
> And can you explain in more depth what retention is still missing. means? You want retention_days in the ORDER BY so that rows with different timestamps can't be merged (even though they should always be partitioned separately)?

Yes, two buckets have different retention_days value but the same values for all the other fields in the order by key, they will be merged together. I don't know what the resulting retention will be. 

Now we are also partitioning by retention, so this should never happen in practice because each partition can only contain one retention value, though that makes the data model less sound, it would be hard to notice the mistake if we change the partitioning and remove retention_days from there.

onewland commented:
Added use_case_id to PRIMARY KEY and ORDER BY
Added retention_days to the end of ORDER BY

SQL:
```
Local operations:
CREATE TABLE IF NOT EXISTS metrics_counters_v2_local (use_case_id LowCardinality(String), org_id UInt64, project_id UInt64, metric_id UInt64, granularity UInt32, tags Nested(key UInt64, value UInt64), timestamp DateTime, retention_days UInt16, value AggregateFunction(sum, Float64)) ENGINE AggregatingMergeTree() PRIMARY KEY (use_case_id, org_id, project_id, metric_id, granularity, timestamp) ORDER BY (use_case_id, org_id, project_id, metric_id, granularity, timestamp, tags.key, tags.value, retention_days) PARTITION BY (retention_days, toMonday(timestamp)) TTL timestamp + toIntervalDay(retention_days) SETTINGS index_granularity=2048;
ALTER TABLE metrics_counters_v2_local ADD COLUMN IF NOT EXISTS _tags_hash Array(UInt64) MATERIALIZED arrayMap((k, v) -> cityHash64(concat(toString(k), '=', toString(v))), tags.key, tags.value) AFTER tags.value;
ALTER TABLE metrics_counters_v2_local ADD INDEX IF NOT EXISTS bf_tags_hash _tags_hash TYPE bloom_filter() GRANULARITY 1;
ALTER TABLE metrics_counters_v2_local ADD INDEX IF NOT EXISTS bf_tags_key_hash tags.key TYPE bloom_filter() GRANULARITY 1;


Dist operations:
Skipped dist operation - single node cluster
Skipped dist operation - single node cluster
```
